File: sdk/js/src/verify.js
// Client-side verification of AgentGraph Trust Score v2 envelopes.
//
// The point of the signed envelope: a consumer can verify a trust score WITHOUT
// trusting AgentGraph's server. This module reproduces the exact verification the
// Python SDK (agentgraph_sdk/verify.py) and the server do — JCS-canonical,
// proof-stripped, Ed25519 over a SHA-256 digest — standalone. It only needs
// `canonicalize` (RFC 8785 JCS, byte-matches Python's rfc8785) plus Node's
// built-in `node:crypto`.
//
// Spec: docs/standards/trust-score-envelope-v2.0.md (§6 verification).

import { createHash, createPublicKey, verify as cryptoVerify } from 'node:crypto';
import canonicalize from 'canonicalize';

export const PROOF_TYPE = 'Ed25519Signature2020';

/**
 * Outcome of verifying a signed trust-score envelope.
 *
 * `valid` is the bottom line (signature good AND within freshness window).
 * The component fields explain a failure.
 *
 * @typedef {Object} VerificationResult
 * @property {boolean} valid           signature valid AND fresh
 * @property {boolean} signatureValid  Ed25519 signature verified
 * @property {boolean} fresh           within freshness_ttl_seconds window
 * @property {string|null} kid         key id read from the JWS header
 * @property {string} reason           human-readable outcome
 */

/** base64url -> Buffer (handles missing padding). */
function b64urlDecode(s) {
  return Buffer.from(s, 'base64url');
}

/** Return a shallow copy of the envelope without the top-level `proof` key. */
function stripProof(envelope) {
  const out = {};
  for (const k of Object.keys(envelope)) {
    if (k !== 'proof') out[k] = envelope[k];
  }
  return out;
}

/**
 * SHA-256 of the JCS-canonical, proof-stripped envelope (what proof.jws signs).
 * Returns the raw 32-byte digest (mirrors Python's `.digest()`, not hexdigest).
 * @param {object} envelope
 * @returns {Buffer}
 */
export function envelopeDigest(envelope) {
  const canonical = canonicalize(stripProof(envelope));
  return createHash('sha256').update(Buffer.from(canonical, 'utf-8')).digest();
}

/** Read the `kid` from a compact JWS header (first dot-part). */
function kidFromJws(jws) {
  try {
    const header = jws.split('.', 1)[0];
    const decoded = JSON.parse(b64urlDecode(header).toString('utf-8'));
    return decoded.kid ?? null;
  } catch {
    return null;
  }
}

/**
 * Resolve an Ed25519 public key (Node KeyObject) from a JWKS (RFC 7517) by kid.
 *
 * `jwks` may be the full `{ keys: [...] }` doc or a bare array of JWKs.
 * If `kid` is null, the first OKP/Ed25519 key is used.
 * @returns {import('node:crypto').KeyObject|null}
 */
function publicKeyForKid(jwks, kid) {
  const keys = Array.isArray(jwks) ? jwks : (jwks?.keys ?? []);
  const candidates = keys.filter(
    (k) => k && k.kty === 'OKP' && k.crv === 'Ed25519' && 'x' in k,
  );
  const build = (jwk) =>
    createPublicKey({
      key: { kty: 'OKP', crv: 'Ed25519', x: jwk.x },
      format: 'jwk',
    });
  if (kid !== null && kid !== undefined) {
    for (const k of candidates) {
      if (k.kid === kid) return build(k);
    }
    return null;
  }
  if (candidates.length > 0) return build(candidates[0]);
  return null;
}

/**
 * True iff the envelope is within its `freshness_ttl_seconds` window.
 * @param {object} envelope
 * @param {{now?: Date}} [opts]
 * @returns {boolean}
 */
export function isFresh(envelope, { now } = {}) {
  const nowDate = now ?? new Date();
  try {
    const computedStr = String(envelope.computed_at).replace('Z', '+00:00');
    const computed = new Date(computedStr);
    if (Number.isNaN(computed.getTime())) return false;
    const ttl = parseInt(envelope.freshness_ttl_seconds ?? 0, 10);
    if (Number.isNaN(ttl)) return false;
    const deltaSeconds = (nowDate.getTime() - computed.getTime()) / 1000;
    return deltaSeconds <= ttl;
  } catch {
    return false;
  }
}

/**
 * Verify a signed v2 trust-score envelope against a JWKS.
 *
 * Fetch the JWKS once from `<issuer>/.well-known/jwks.json` (or use the client's
 * `verifyEnvelope()`/`verify()` which do it for you). Returns a
 * VerificationResult; `valid` is true iff signature valid AND fresh.
 *
 * @param {object} envelope
 * @param {object|Array} jwks
 * @param {{now?: Date}} [opts]
 * @returns {VerificationResult}
 */
export function verifyEnvelope(envelope, jwks, { now } = {}) {
  const proof = envelope?.proof;
  if (!proof || proof.type !== PROOF_TYPE) {
    return { valid: false, signatureValid: false, fresh: false, kid: null, reason: 'missing or unsupported proof' };
  }
  const jws = proof.jws ?? '';
  const parts = jws.split('.');
  if (parts.length !== 3) {
    return { valid: false, signatureValid: false, fresh: false, kid: null, reason: 'malformed jws' };
  }

  const kid = kidFromJws(jws);
  const pub = publicKeyForKid(jwks, kid);
  if (pub === null) {
    return { valid: false, signatureValid: false, fresh: false, kid, reason: 'no matching key in JWKS' };
  }

  let sigOk = false;
  try {
    const signature = b64urlDecode(parts[2]);
    // Ed25519: algorithm MUST be null in Node's crypto.verify.
    sigOk = cryptoVerify(null, envelopeDigest(envelope), pub, signature);
  } catch {
    sigOk = false;
  }

  const fresh = isFresh(envelope, { now });
  if (!sigOk) {
    return { valid: false, signatureValid: false, fresh, kid, reason: 'signature invalid' };
  }
  if (!fresh) {
    return { valid: false, signatureValid: true, fresh: false, kid, reason: 'envelope expired (stale)' };
  }
  return { valid: true, signatureValid: true, fresh: true, kid, reason: 'ok' };
}

export default verifyEnvelope;

File: sdk/agentgraph_sdk/verify.py
"""Client-side verification of AgentGraph Trust Score v2 envelopes.

The point of the signed envelope: a consumer can verify a trust score WITHOUT
trusting AgentGraph's server. This module reproduces the exact verification the
server does (JCS-canonical, proof-stripped, Ed25519 over a SHA-256 digest),
standalone — it only needs ``rfc8785`` + ``cryptography``, not any AgentGraph
server code.

Spec: docs/standards/trust-score-envelope-v2.0.md (§6 verification).
"""
from __future__ import annotations

import base64
import hashlib
from dataclasses import dataclass
from datetime import datetime, timezone

import rfc8785
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

PROOF_TYPE = "Ed25519Signature2020"


@dataclass(frozen=True)
class VerificationResult:
    """Outcome of verifying a signed trust-score envelope.

    ``valid`` is the bottom line (signature good AND within freshness window).
    The component fields explain a failure.
    """

    valid: bool
    signature_valid: bool
    fresh: bool
    kid: str | None = None
    reason: str = "ok"

    def __bool__(self) -> bool:
        return self.valid


def _b64url_decode(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def _strip_proof(envelope: dict) -> dict:
    return {k: v for k, v in envelope.items() if k != "proof"}


def envelope_digest(envelope: dict) -> bytes:
    """SHA-256 of the JCS-canonical, proof-stripped envelope (what proof.jws signs)."""
    return hashlib.sha256(rfc8785.dumps(_strip_proof(envelope))).digest()


def _kid_from_jws(jws: str) -> str | None:
    try:
        header = jws.split(".", 1)[0]
        import json

        return json.loads(_b64url_decode(header)).get("kid")
    except Exception:
        return None


def _public_key_for_kid(jwks: dict | list, kid: str | None) -> Ed25519PublicKey | None:
    """Resolve an Ed25519 public key from a JWKS (RFC 7517) by kid.

    ``jwks`` may be the full ``{"keys": [...]}`` doc or a bare list of JWKs.
    If ``kid`` is None, the first OKP/Ed25519 key is used.
    """
    keys = jwks.get("keys", []) if isinstance(jwks, dict) else jwks
    candidates = [
        k for k in keys
        if k.get("kty") == "OKP" and k.get("crv") == "Ed25519" and "x" in k
    ]
    if kid is not None:
        for k in candidates:
            if k.get("kid") == kid:
                return Ed25519PublicKey.from_public_bytes(_b64url_decode(k["x"]))
        return None
    if candidates:
        return Ed25519PublicKey.from_public_bytes(_b64url_decode(candidates[0]["x"]))
    return None


def is_fresh(envelope: dict, *, now: datetime | None = None) -> bool:
    """True iff the envelope is within its ``freshness_ttl_seconds`` window."""
    now = now or datetime.now(timezone.utc)
    try:
        computed = datetime.fromisoformat(envelope["computed_at"].replace("Z", "+00:00"))
        ttl = int(envelope.get("freshness_ttl_seconds", 0))
    except (KeyError, ValueError):
        return False
    return (now - computed).total_seconds() <= ttl


def verify_envelope(
    envelope: dict, jwks: dict | list, *, now: datetime | None = None
) -> VerificationResult:
    """Verify a signed v2 trust-score envelope against a JWKS.

    Fetch the JWKS once from ``<issuer>/.well-known/jwks.json`` (or use the
    client's ``verify()`` which does it for you). Returns a
    :class:`VerificationResult`; truthy iff signature valid AND fresh.
    """
    proof = envelope.get("proof")
    if not proof or proof.get("type") != PROOF_TYPE:
        return VerificationResult(False, False, False, reason="missing or unsupported proof")
    jws = proof.get("jws", "")
    parts = jws.split(".")
    if len(parts) != 3:
        return VerificationResult(False, False, False, reason="malformed jws")

    kid = _kid_from_jws(jws)
    pub = _public_key_for_kid(jwks, kid)
    if pub is None:
        return VerificationResult(False, False, False, kid=kid, reason="no matching key in JWKS")

    try:
        pub.verify(_b64url_decode(parts[2]), envelope_digest(envelope))
        sig_ok = True
    except (InvalidSignature, ValueError):
        sig_ok = False

    fresh = is_fresh(envelope, now=now)
    if not sig_ok:
        return VerificationResult(False, False, fresh, kid=kid, reason="signature invalid")
    if not fresh:
        return VerificationResult(False, True, False, kid=kid, reason="envelope expired (stale)")
    return VerificationResult(True, True, True, kid=kid, reason="ok")


__all__ = ["VerificationResult", "verify_envelope", "is_fresh", "envelope_digest"]
